doc/architecture/04solution.md

```diff
@@ -16,9 +16,11 @@ We use specific types (prefer enum over int over string etc.) and share types wh
 ## Validation
 
 - The frontend should validate data as early as possible, usually during input using [React Hook Form](https://react-hook-form.com/).
-- Important data should be also validated in the backend.
-  Type validation is done by actix web, but some constraints could be bypassed by just validating in the frontend.
-  For example, payment data or data associated with gamification achievements.
+- For more complex validation logic, we use [Zod](https://zod.dev/) in conjunction with [React Hook Form Schema Validation](https://react-hook-form.com/get-started#SchemaValidation).
+  For example Zod should be used over default React Hook Form validation when there are inter-dependent form values that need to be validated together, such as validating that a start date is before an end date.
+- Important data should be validated in the backend, even if type validation is already done by actix web.
```

Comment:
All data should be validatet in the backend, even if type validation is already done by actix web.

Reply:
What I meant by type validation is validation that the type is correct, e.g. it can be deserialized from the request body.
From the actix web docs: https://docs.rs/actix-web/latest/actix_web/web/struct.Json.html#extractor
It uses serde underneath.

Reply:
@markus2330 @temmey 
How about:
All constraints that go beyond the type level validation should be explicitly validated by the backend as well as the frontend.

Reply:
From a security perspective, the backend should be enough. Since whatever kind of validation we implement in the frontend, it can be bypassed by a malicious user.

Reply:
> From the actix web docs: https://docs.rs/actix-web/latest/actix_web/web/struct.Json.html#extractor
It uses serde underneath.
> All constraints that go beyond the type level validation should be explicitly validated by the backend as well as the frontend.

Great input, I tried to squeeze everything in one sentence:

```suggestion
- All constraints that go beyond the types, e.g. automated validation while [extracting via serde](https://docs.rs/actix-web/latest/actix_web/web/struct.Json.html#extractor), should be explicitly validated by the backend as well as the frontend.
```